libp2p/nameresolving/dnsresolver.nim

```diff
@@ -0,0 +1,147 @@
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+{.push raises: [Defect].}
+
+import
+  std/[streams, strutils, sets, sequtils],
+  chronos, chronicles,
+  dnsclientpkg/[protocol, types]
+
+import
+  nameresolver
+
+logScope:
+  topics = "libp2p dnsresolver"
+
+type
+  DnsResolver* = ref object of NameResolver
+    nameServers*: seq[TransportAddress]
+
+proc questionToBuf(address: string, kind: QKind): seq[byte] =
+  try:
+    var
+      header = initHeader()
+      question = initQuestion(address, kind)
+
+      requestStream = header.toStream()
+    question.toStream(requestStream)
+
+    let dataLen = requestStream.getPosition()
+    requestStream.setPosition(0)
+
+    var buf = newSeq[byte](dataLen)
+    discard requestStream.readData(addr buf[0], dataLen)
+    return buf
+  except CatchableError as exc:
+    info "Failed to created DNS buffer", msg = exc.msg
+    return newSeq[byte](0)
+
+proc getDnsResponse(
+  dnsServer: TransportAddress,
+  address: string,
+  kind: QKind): Future[Response] {.async.} =
+
+  var sendBuf = questionToBuf(address, kind)
+
+  if sendBuf.len == 0:
+    raise newException(ValueError, "Incorrect DNS query")
+
+  let receivedDataFuture = newFuture[void]()
+
+  proc datagramDataReceived(transp: DatagramTransport,
+                  raddr: TransportAddress): Future[void] {.async, closure.} =
+      receivedDataFuture.complete()
+
+  let sock =
+    if dnsServer.family == AddressFamily.IPv6:
+      newDatagramTransport6(datagramDataReceived)
+    else:
+      newDatagramTransport(datagramDataReceived)
+
+  try:
+    await sock.sendTo(dnsServer, addr sendBuf[0], sendBuf.len)
+
+    await receivedDataFuture or sleepAsync(5.seconds) #unix default
+
+    if not receivedDataFuture.finished:
+      raise newException(ValueError, "DNS server timeout")
+
+    var
+      rawResponse = sock.getMessage()
+      dataStream = newStringStream()
+    dataStream.writeData(addr rawResponse[0], rawResponse.len)
+    dataStream.setPosition(0)
+    return parseResponse(dataStream)
+  finally:
+    await sock.closeWait()
+
+method resolveIp*(
+  self: DnsResolver,
+  address: string,
+  port: Port,
+  domain: Domain = Domain.AF_UNSPEC): Future[seq[TransportAddress]] {.async.} =
+
+  let nameservers = self.nameServers
+  trace "Resolving IP using DNS", address, servers = nameservers.mapIt($it), domain
+  for server in nameservers:
+    var responseFutures: seq[Future[Response]]
+    if domain == Domain.AF_INET or domain == Domain.AF_UNSPEC:
+      responseFutures.add(getDnsResponse(server, address, A))
+
+    if domain == Domain.AF_INET6 or domain == Domain.AF_UNSPEC:
+      let fut = getDnsResponse(server, address, AAAA)
+      if server.family == AddressFamily.IPv6:
+        trace "IPv6 DNS server, puting AAAA records first", server = $server
+        responseFutures.insert(fut)
```

Comment:
I'm curious, do other resolvers also give preference to `ipv6` addrs?

Reply:
Really tricky topic, it depends a lot on os etc.

What, eg, browsers will do is connect both in ipv4 & ipv6, and if ipv6 succeeds, kill the ipv4 connection. (To give preference to ipv6, but also have a quick fallback if the connection failed). (as per the Happy Eyeball RFC)
The cost is obviously more resources usages, since you need to create two sockets everytime you connect to a new server, at least, but you get better latency in case of non-working ipv6 (on the client or server side)

I think in our case, I think that would be ill-advised, we should keep the resources usages to a minimum.
So, if we succeed to retrieve the addresses from a IPV6 DNS server, we'll put the IPV6 resolved addresses first, since we know that at least the client has a correct ipv6 configuration.

If we end up in a latency constrained situation, we should instead use a different dialer (for instance a "quick dial"), which tries multiple `ma` in parallel to be sure that we have a working connection ASAP

Though I'll improve a bit things here, by "saving" the fact that a DNS server failed to resolve for future resolves.

Reply:
What concerns me is the success ratio of Ipv6 vs Ipv4 dials. If Ipv6 fails to connect in a large amount of cases (not unreasonable, given it's uneven adoption), then the net effect would be much slower dials.

> If we end up in a latency constrained situation, we should instead use a different dialer (for instance a "quick dial"), which tries multiple ma in parallel to be sure that we have a working connection ASAP

Hmm, not sure, I don't have a strong opinion here, but other implementations of libp2p default to concurrent dials, like you note, this has advantages and disadvantages. My plan has been to add parallel/concurrent dials by default at some point, but you might be right in that this should be user configurable and not enforced.